### PR TITLE
Support and use broadcast with mapreduce.

### PR DESCRIPTION
**Pull request by maleadt**

### CODE REVIEW

**Comment by maleadt** [comment]
```julia
julia> mapreduce(+,+,jl(ones(1)), jl(ones(2,2)))
8.0

julia> mapreduce(+,+,ones(1), ones(2,2))
2.0
```

Looks like we can't implement the implicit `map(+, ..., ...)` here using Broadcast?

```julia
julia> map(+, ones(1), ones(2,2))
1-element Array{Float64,1}:
 2.0

julia> ones(1) .+ ones(2,2)
2×2 Array{Float64,2}:
 2.0  2.0
 2.0  2.0
```

File: src/host/mapreduce.jl [hunk 1]
```diff
@@ -1,9 +1,14 @@
 # map-reduce
 
+const AbstractArrayOrBroadcasted = Union{AbstractArray,Broadcast.Broadcasted}
+
 # GPUArrays' mapreduce methods build on `Base.mapreducedim!`, but with an additional
 # argument `init` value to avoid eager initialization of `R` (if set to something).
-mapreducedim!(f, op, R::AbstractGPUArray, As::AbstractArray...; init=nothing) = error("Not implemented") # COV_EXCL_LINE
+mapreducedim!(f, op, R::AbstractGPUArray, A::AbstractArrayOrBroadcasted;
+              init=nothing) = error("Not implemented") # COV_EXCL_LINE
+# resolve ambiguities
```
Comment (maleadt) [inline src/host/mapreduce.jl:9]: @tkf I guess you didn't run into such ambiguities with https://github.com/JuliaLang/julia/pull/31020?

Comment (maleadt) [reply src/host/mapreduce.jl:9]: e.g., for the `mapreduce` case below:

```
ERROR: MethodError: mapreduce(::typeof(+), ::typeof(+), ::JLArray{Float64,1}, ::JLArray{Float64,2}) is ambiguous. Candidates:
  mapreduce(f, op, A::AbstractArray...; kw...) in Base at reducedim.jl:308
  mapreduce(f, op, A::Union{Base.Broadcast.Broadcasted, AbstractArray}, As::Union{Base.Broadcast.Broadcasted, AbstractArray}...; dims, init) in GPUArrays at /home/tim/Julia/pkg/GPUArrays/src/host/mapreduce.jl:27
Possible fix, define
  mapreduce(::Any, ::Any, ::AbstractArray, ::Vararg{AbstractArray,N} where N)
```

Comment (tkf) [reply src/host/mapreduce.jl:9]: I don't remember seeing anything similar to this but it's a long time ago that I wrote that patch.

**Pull request merged**